geopandas/io/sql.py

```diff
@@ -47,17 +48,30 @@ def read_postgis(sql, con, geom_col='geom', crs=None, hex_encoded=True,
     if geom_col not in df:
         raise ValueError("Query missing geometry column '{}'".format(geom_col))
 
-    def load_geom(x):
-        if isinstance(x, bytes):
-            return shapely.wkb.loads(x, hex=hex_encoded)
-        else:
+    geoms = df[geom_col].dropna()
+
+    if not geoms.empty:
+        load_geom_bytes = shapely.wkb.loads
+
+        def load_geom_buffer(x):
+            return shapely.wkb.loads(str(x))
+
+        def load_geom_str(x):
             return shapely.wkb.loads(str(x), hex=hex_encoded)
-    geoms = df[geom_col].apply(load_geom)
-    df[geom_col] = geoms
 
-    if crs is None:
-        if len(geoms) > 0:
-            srid = shapely.geos.lgeos.GEOSGetSRID(geoms[0]._geom)
+        if sys.version_info.major < 3:
+            if isinstance(geoms.iat[0], buffer):
+                load_geom = load_geom_buffer
+            else:
+                load_geom = load_geom_str
+        elif isinstance(geoms.iat[0], bytes):
+            load_geom = load_geom_bytes
+        else:
+            load_geom = load_geom_str
```

Comment:
Thanks for explaining the rationale of the original writing of this, as well as re-writing it like this (which, I think, does help).